--- a/net/mac80211/tx.c	2026-06-22 14:18:55.885184669 +0800
+++ b/net/mac80211/tx.c	2026-06-22 18:55:22.284756054 +0800
@@ -3807,6 +3807,7 @@
 	struct ieee80211_vif *vif = txq->vif;
 	int q = vif->hw_queue[txq->ac];
 	bool q_stopped;
+	unsigned long last_qs = jiffies;
 
 	WARN_ON_ONCE(softirq_count() == 0);
 
@@ -3846,6 +3847,8 @@
 
 	spin_unlock_bh(&fq->lock);
 
+	rcu_softirq_qs_periodic(last_qs);
+
 	hdr = (struct ieee80211_hdr *)skb->data;
 	info = IEEE80211_SKB_CB(skb);
 
