--- a/drivers/irqchip/irq-mips-gic.c	2024-01-26 06:34:33.000000000 +0800
+++ b/drivers/irqchip/irq-mips-gic.c	2026-07-18 17:19:25.260162642 +0800
@@ -176,21 +176,28 @@
 static void gic_mask_irq(struct irq_data *d)
 {
 	unsigned int intr = GIC_HWIRQ_TO_SHARED(d->hwirq);
+	unsigned long flags;
 
 	write_gic_rmask(intr);
+
+	raw_spin_lock_irqsave(&gic_lock, flags);
 	gic_clear_pcpu_masks(intr);
+	raw_spin_unlock_irqrestore(&gic_lock, flags);
 }
 
 static void gic_unmask_irq(struct irq_data *d)
 {
 	unsigned int intr = GIC_HWIRQ_TO_SHARED(d->hwirq);
 	unsigned int cpu;
+	unsigned long flags;
 
 	write_gic_smask(intr);
 
+	raw_spin_lock_irqsave(&gic_lock, flags);
 	gic_clear_pcpu_masks(intr);
 	cpu = cpumask_first(irq_data_get_effective_affinity_mask(d));
 	set_bit(intr, per_cpu_ptr(pcpu_masks, cpu));
+	raw_spin_unlock_irqrestore(&gic_lock, flags);
 }
 
 static void gic_ack_irq(struct irq_data *d)
