--- a/mt7915/mmio.c
+++ b/mt7915/mmio.c
@@ -638,16 +638,22 @@
 		int token;
 		void *ptr;
 
+		if (!t)
+			goto unmap;
+
 		ptr = page_frag_alloc(&wed->rx_buf_ring.rx_page, length,
 				      GFP_KERNEL);
-		if (!ptr)
+		if (!ptr) {
+			mt76_put_rxwi(&dev->mt76, t);
 			goto unmap;
+		}
 
 		phy_addr = dma_map_single(dev->mt76.dma_dev, ptr,
 					  wed->wlan.rx_size,
 					  DMA_TO_DEVICE);
 		if (unlikely(dma_mapping_error(dev->mt76.dev, phy_addr))) {
 			skb_free_frag(ptr);
+			mt76_put_rxwi(&dev->mt76, t);
 			goto unmap;
 		}
 
@@ -657,6 +663,7 @@
 			dma_unmap_single(dev->mt76.dma_dev, phy_addr,
 					 wed->wlan.rx_size, DMA_TO_DEVICE);
 			skb_free_frag(ptr);
+			mt76_put_rxwi(&dev->mt76, t);
 			goto unmap;
 		}
 
@@ -716,6 +723,8 @@
 		wed->wlan.base = devm_ioremap(dev->mt76.dev,
 					      pci_resource_start(pci_dev, 0),
 					      pci_resource_len(pci_dev, 0));
+		if (!wed->wlan.base)
+			return -ENOMEM;
 		wed->wlan.phy_base = pci_resource_start(pci_dev, 0);
 		wed->wlan.wpdma_int = pci_resource_start(pci_dev, 0) +
 				      MT_INT_WED_SOURCE_CSR;
@@ -743,6 +752,8 @@
 		wed->wlan.bus_type = MTK_WED_BUS_AXI;
 		wed->wlan.base = devm_ioremap(dev->mt76.dev, res->start,
 					      resource_size(res));
+		if (!wed->wlan.base)
+			return -ENOMEM;
 		wed->wlan.phy_base = res->start;
 		wed->wlan.wpdma_int = res->start + MT_INT_SOURCE_CSR;
 		wed->wlan.wpdma_mask = res->start + MT_INT_MASK_CSR;
