--- a/mt7915/mt7915.h
+++ b/mt7915/mt7915.h
@@ -334,6 +334,8 @@
 		bool restart:1;
 	} recovery;
 
+	u32 mcu_poll_cnt;
+
 	/* protects coredump data */
 	struct mutex dump_mutex;
 #ifdef CONFIG_DEV_COREDUMP
--- a/mt7915/dma.c
+++ b/mt7915/dma.c
@@ -28,8 +28,16 @@
 	struct mt7915_dev *dev;
 
 	dev = container_of(napi, struct mt7915_dev, mt76.tx_napi);
-
+	
 	mt76_connac_tx_cleanup(&dev->mt76);
+	
+	if (++dev->mcu_poll_cnt < 16) {
+		napi_complete(napi);
+        	napi_schedule(napi);
+        	return 0;
+	}
+	dev->mcu_poll_cnt = 0;
+
 	if (napi_complete_done(napi, 0))
 		mt7915_irq_enable(dev, MT_INT_TX_DONE_MCU);
 
@@ -388,6 +396,7 @@
 	u32 hif1_ofs = 0;
 	int ret;
 
+	dev->mcu_poll_cnt = 0; // used for mt7915_poll_tx
 	mt7915_dma_config(dev);
 
 	mt76_dma_attach(&dev->mt76);
