--- a/mt7915/init.c
+++ b/mt7915/init.c
@@ -405,7 +405,7 @@
 			IEEE80211_HT_CAP_LDPC_CODING |
 			IEEE80211_HT_CAP_MAX_AMSDU;
 		phy->mt76->sband_2g.sband.ht_cap.ampdu_density =
-			IEEE80211_HT_MPDU_DENSITY_4;
+			IEEE80211_HT_MPDU_DENSITY_8;
 	}
 
 	if (phy->mt76->cap.has_5ghz) {
@@ -416,7 +416,7 @@
 			IEEE80211_HT_CAP_LDPC_CODING |
 			IEEE80211_HT_CAP_MAX_AMSDU;
 		phy->mt76->sband_5g.sband.ht_cap.ampdu_density =
-			IEEE80211_HT_MPDU_DENSITY_4;
+			IEEE80211_HT_MPDU_DENSITY_8;
 
 		if (is_mt7915(&dev->mt76)) {
 			vht_cap->cap |=
