--- a/mt7915/mac.c
+++ b/mt7915/mac.c
@@ -2114,11 +2114,13 @@
 {
 	struct mt7915_phy *phy;
 	struct mt76_phy *mphy;
+	struct mt7915_dev *dev;
 	bool clear = false; 
 
 	mphy = (struct mt76_phy *)container_of(work, struct mt76_phy,
 					       mac_work.work);
 	phy = mphy->priv;
+	dev = phy->dev;
 
 	mutex_lock(&mphy->dev->mutex);
 
@@ -2141,9 +2143,21 @@
 
 	if (clear) {
 		u32 i;
-		mt7915_mac_sta_poll_clear(phy->dev); // Reset sta ADM Counts
+		mt7915_mac_sta_poll_clear(dev); // Reset sta ADM Counts
 		for (i = 0; i < 2; i++) // Reset global Band (backoff/eifs/obss)
-			mt7915_mac_refresh_wash(phy->dev, i);
+			mt7915_mac_refresh_wash(dev, i);
+		mt7915_mac_cca_stats_reset(phy); // CCA reset
+		// Counters reset (AGG M1/M2.. reset)
+		for (i = 0; i < 4; i++) {
+                	mt76_rr(dev, MT_TX_AGG_CNT(phy->mt76->band_idx, i));
+                	mt76_rr(dev, MT_TX_AGG_CNT2(phy->mt76->band_idx, i));
+        	}
+        	memset(phy->mt76->aggr_stats, 0, sizeof(phy->mt76->aggr_stats));
+
+	        /* reset airtime counters */
+        	mt76_set(dev, MT_WF_RMAC_MIB_AIRTIME0(phy->mt76->band_idx),
+                	MT_WF_RMAC_MIB_RXTIME_CLR);
+
 	}
 
 	ieee80211_queue_delayed_work(mphy->hw, &mphy->mac_work,
