--- a/mt76.h
+++ b/mt76.h
@@ -741,6 +741,9 @@
 	} leds;
 };
 
+extern atomic_t rx_poll_retry_cnt;
+extern int rx_poll_timeo;
+
 struct mt76_dev {
 	struct mt76_phy phy; /* must be first */
 	struct mt76_phy *phys[__MT_MAX_BAND];
--- a/debugfs.c
+++ b/debugfs.c
@@ -122,6 +122,39 @@
 DEFINE_DEBUGFS_ATTRIBUTE(fops_rx_nframes_limit, mt76_rx_nframes_limit_get, mt76_rx_nframes_limit_set,
                          "%lld\n");
 
+static int
+mt76_rx_poll_retry_cnt_read(struct seq_file *s, void *data)
+{
+	//struct mt76_dev *dev = dev_get_drvdata(s->private);
+
+        seq_printf(s, "rx_poll_retry_cnt: %d\n", atomic_read(&rx_poll_retry_cnt));
+
+        return 0;
+}
+
+static int
+mt76_rx_poll_timeo_set(void *data, u64 val)
+{
+        //struct mt76_dev *dev = data;
+
+        rx_poll_timeo = (int)val;
+
+        return 0;
+}
+
+static int
+mt76_rx_poll_timeo_get(void *data, u64 *val)
+{
+        //struct mt76_dev *dev = data;
+
+        *val = (u64)rx_poll_timeo;
+
+        return 0;
+}
+
+DEFINE_DEBUGFS_ATTRIBUTE(fops_rx_poll_timeo, mt76_rx_poll_timeo_get, mt76_rx_poll_timeo_set,
+                         "%lld\n");
+
 
 void mt76_seq_puts_array(struct seq_file *file, const char *str,
 			 s8 *val, int len)
@@ -159,6 +192,9 @@
 	debugfs_create_devm_seqfile(dev->dev, "rx-aggr-nframes", dir, mt76_rx_nframes_read);
 	debugfs_create_file_unsafe("rx-aggr-nframes-limit", 0600, dir, dev, &fops_rx_nframes_limit);
 
+	debugfs_create_devm_seqfile(dev->dev, "rx-poll-retry-cnt", dir, mt76_rx_poll_retry_cnt_read);
+	debugfs_create_file_unsafe("rx-poll-timeo", 0600, dir, dev, &fops_rx_poll_timeo);
+
 	return dir;
 }
 EXPORT_SYMBOL_GPL(mt76_register_debugfs_fops);
--- a/dma.c
+++ b/dma.c
@@ -873,14 +873,66 @@
 	return done;
 }
 
+#define MAX_RX_POLL_RETRY 3
+#define DEFAULT_RX_POLL_TIMEO 1000 // us
+#define MAX_SAMPLE_COUNT 100
+atomic_t rx_poll_retry_cnt = ATOMIC_INIT(MAX_RX_POLL_RETRY);
+int rx_poll_timeo = DEFAULT_RX_POLL_TIMEO;
+
+static void mt76_rx_poll_retry_adjust(u64 start_jif)
+{
+	static struct {
+		u32 total_elapsed;
+		u16 sample_cnt;
+		u16 total_sample_cnt;
+		u8 max_retry_sum;
+	} ____cacheline_aligned rx_stats;
+	u64 elapsed = ktime_get() - start_jif;
+	int val = atomic_read(&rx_poll_retry_cnt);
+
+	if (rx_stats.sample_cnt++ < MAX_SAMPLE_COUNT) {
+		rx_stats.total_elapsed += elapsed;
+		return;
+	}
+	rx_stats.total_sample_cnt += rx_stats.sample_cnt;
+
+	if (rx_stats.total_elapsed >= rx_poll_timeo*1000*MAX_SAMPLE_COUNT) {
+		if (val > 1) {
+			atomic_dec(&rx_poll_retry_cnt);
+		}
+	} else {
+		if (val < MAX_RX_POLL_RETRY)
+			atomic_inc(&rx_poll_retry_cnt);
+	}
+
+	rx_stats.sample_cnt = 0; rx_stats.total_elapsed = 0; // reset
+
+	// check if the accumated count at MAX_RX_POLL_RETRY
+	val = atomic_read(&rx_poll_retry_cnt);
+	if (val == MAX_RX_POLL_RETRY)
+		rx_stats.max_retry_sum++; // recording the max_retry_sum
+	if (rx_stats.total_sample_cnt >= MAX_SAMPLE_COUNT*10) {
+		if (rx_stats.max_retry_sum >= 7) {
+			rx_poll_timeo -= 100; // us
+		} else {
+			rx_poll_timeo += 100; // us
+		}
+		if (rx_poll_timeo < 0) rx_poll_timeo = 0; // avoiding integer overflow when comparing to u32 or u64
+		rx_stats.total_sample_cnt = 0; rx_stats.max_retry_sum = 0; // reset
+	}
+}
+
 int mt76_dma_rx_poll(struct napi_struct *napi, int budget)
 {
-	struct mt76_dev *dev;
+	struct mt76_dev *dev = NULL;
 	int qid, done = 0, cur;
+	int retry_limit = atomic_read(&rx_poll_retry_cnt); 
+	u64 start_jif = ktime_get();
 
 	dev = container_of(napi->dev, struct mt76_dev, napi_dev);
 	qid = napi - dev->napi;
 
+rev_dma:
 	rcu_read_lock();
 
 	do {
@@ -894,8 +946,19 @@
 	if (done < budget && napi_complete(napi)) {
 		if (done) wmb(); // sync. cpu write
 		dev->drv->rx_poll_complete(dev, qid);
+		mt76_rx_poll_retry_adjust(start_jif);
+		return done;
+	}else if (--retry_limit > 0) {
+		if (unlikely(tif_need_resched())) {
+			cond_resched();
+			mt76_rx_poll_retry_adjust(start_jif);
+			return done;
+		}
+		done = 0;
+		goto rev_dma;
 	}
 
+	mt76_rx_poll_retry_adjust(start_jif);
 	return done;
 }
 EXPORT_SYMBOL_GPL(mt76_dma_rx_poll);
