--- a/dma.c
+++ b/dma.c
@@ -877,10 +877,12 @@
 {
 	struct mt76_dev *dev;
 	int qid, done = 0, cur;
+	int retry_limit = 3; // greedy recieve skb from DMA
 
 	dev = container_of(napi->dev, struct mt76_dev, napi_dev);
 	qid = napi - dev->napi;
 
+rev_dma:
 	rcu_read_lock();
 
 	do {
@@ -894,6 +896,14 @@
 	if (done < budget && napi_complete(napi)) {
 		if (done) wmb(); // sync. cpu write
 		dev->drv->rx_poll_complete(dev, qid);
+		return done;
+	}else if (--retry_limit > 0) {
+		if (unlikely(tif_need_resched())) {
+			cond_resched();
+			return done;
+		}
+		done = 0;
+		goto rev_dma;
 	}
 
 	return done;
