--- a/mac80211.c	2026-07-07 14:35:40.836815265 +0800
+++ b/mac80211.c	2026-07-07 14:34:56.311586404 +0800
@@ -923,17 +923,24 @@
 		if (dev->drv->drv_flags & MT_DRV_SW_RX_AIRTIME)
 			survey->filled |= SURVEY_INFO_TIME_BSS_RX;
 	}
-
-	survey->time_busy = div_u64(state->cc_busy, 1000);
-	survey->time_rx = div_u64(state->cc_rx, 1000);
-	survey->time = div_u64(state->cc_active, 1000);
+#define ROUND_DIV_1000(_val) ({ \
+	u64 _x = (u64)(_val) + 500ULL; \
+	(u64)((_x * 42949673ULL) >> 32); \
+})
+	survey->time_busy = ROUND_DIV_1000(state->cc_busy);
+	survey->time_rx = ROUND_DIV_1000(state->cc_rx);
+	survey->time = ROUND_DIV_1000(state->cc_active);
 	survey->noise = state->noise;
 
-	spin_lock_bh(&dev->cc_lock);
-	survey->time_bss_rx = div_u64(state->cc_bss_rx, 1000);
-	survey->time_tx = div_u64(state->cc_tx, 1000);
-	spin_unlock_bh(&dev->cc_lock);
-
+	if (dev->drv->drv_flags & MT_DRV_SW_RX_AIRTIME) {
+		spin_lock_bh(&dev->cc_lock);
+		survey->time_bss_rx = ROUND_DIV_1000(state->cc_bss_rx);
+		survey->time_tx = ROUND_DIV_1000(state->cc_tx);
+		spin_unlock_bh(&dev->cc_lock);
+	} else { // MIB query from firmware
+                survey->time_bss_rx = ROUND_DIV_1000(state->cc_bss_rx);
+                survey->time_tx = ROUND_DIV_1000(state->cc_tx);
+	}
 out:
 	mutex_unlock(&dev->mutex);
 
