--- a/mac80211.c	2026-07-06 08:30:13.032903074 +0800
+++ b/mac80211.c	2026-07-06 09:05:50.381480187 +0800
@@ -912,18 +912,19 @@
 
 	memset(survey, 0, sizeof(*survey));
 	survey->channel = chan;
-	survey->filled = SURVEY_INFO_TIME | SURVEY_INFO_TIME_BUSY;
-	survey->filled |= dev->drv->survey_flags;
+//	survey->filled = SURVEY_INFO_TIME | SURVEY_INFO_TIME_BUSY;
+//	survey->filled |= dev->drv->survey_flags;
+	survey->filled = 0;
 	if (state->noise)
 		survey->filled |= SURVEY_INFO_NOISE_DBM;
 
 	if (chan == phy->main_chan) {
 		survey->filled |= SURVEY_INFO_IN_USE;
 
-		if (dev->drv->drv_flags & MT_DRV_SW_RX_AIRTIME)
-			survey->filled |= SURVEY_INFO_TIME_BSS_RX;
+//		if (dev->drv->drv_flags & MT_DRV_SW_RX_AIRTIME)
+//			survey->filled |= SURVEY_INFO_TIME_BSS_RX;
 	}
-
+/*
 	survey->time_busy = div_u64(state->cc_busy, 1000);
 	survey->time_rx = div_u64(state->cc_rx, 1000);
 	survey->time = div_u64(state->cc_active, 1000);
@@ -933,7 +934,7 @@
 	survey->time_bss_rx = div_u64(state->cc_bss_rx, 1000);
 	survey->time_tx = div_u64(state->cc_tx, 1000);
 	spin_unlock_bh(&dev->cc_lock);
-
+*/
 out:
 	mutex_unlock(&dev->mutex);
 
