--- a/mt76.h	2026-06-27 22:32:12.661558489 +0800
+++ b/mt76.h	2026-06-27 22:01:38.156036796 +0800
@@ -223,6 +223,7 @@
 	int (*mcu_rd_rp)(struct mt76_dev *dev, u32 base,
 			 struct mt76_reg_pair *rp, int len);
 	int (*mcu_restart)(struct mt76_dev *dev);
+	void (*mcu_set_timeout)(struct mt76_dev *mdev, int cmd);
 };
 
 struct mt76_queue_ops {
@@ -524,6 +525,7 @@
 	struct mutex mutex;
 	u32 msg_seq;
 	int timeout;
+        unsigned long mib_access_time;
 
 	struct sk_buff_head res_q;
 	wait_queue_head_t wait;
@@ -1390,7 +1392,9 @@
 
 void mt76_mcu_rx_event(struct mt76_dev *dev, struct sk_buff *skb);
 struct sk_buff *mt76_mcu_get_response(struct mt76_dev *dev,
-				      unsigned long expires);
+				      unsigned long expires, int cmd);
+bool mt76_get_test_mcu_restart(void);
+void mt76_set_test_mcu_restart(bool ok);
 int mt76_mcu_send_and_get_msg(struct mt76_dev *dev, int cmd, const void *data,
 			      int len, bool wait_resp, struct sk_buff **ret);
 int mt76_mcu_skb_send_and_get_msg(struct mt76_dev *dev, struct sk_buff *skb,
--- a/mcu.c	2026-06-27 22:32:12.609556292 +0800
+++ b/mcu.c	2026-06-27 22:31:14.787112875 +0800
@@ -3,7 +3,70 @@
  * Copyright (C) 2019 Lorenzo Bianconi <lorenzo.bianconi83@gmail.com>
  */
 
+#include <linux/wait.h>
+#include <linux/rcupdate.h>
 #include "mt76.h"
+#include "mt76_connac_mcu.h"
+
+/*
+#define mt76_busy_wait_event_timeout(wq_head, condition, timeout)	\
+do {									\
+	unsigned long __last_qs = jiffies;				\
+	unsigned long __end = __last_qs + (timeout);			\
+	for (;;) {							\
+		rmb();							\
+		if (condition)						\
+			break;						\
+		if (time_after_eq(jiffies, __end))			\
+			break;						\
+		cond_resched();						\
+		msleep(1); cpu_relax();				\
+		rcu_softirq_qs_periodic(__last_qs);			\
+	}								\
+} while (0)
+*/
+
+#define ___mt76_wait_event(wq_head, condition, state, exclusive, ret, cmd)      \
+({                                                                              \
+        __label__ __out;                                                        \
+        struct wait_queue_entry __wq_entry;                                     \
+        long __ret = ret;       /* explicit shadow */                           \
+	unsigned long __end = jiffies + (ret);					\
+	unsigned int wq_flags = (exclusive || (cmd == MCU_EXT_CMD(GET_MIB_INFO))) ? WQ_FLAG_EXCLUSIVE : 0; \
+										\
+        init_wait_entry(&__wq_entry, wq_flags);					\
+        for (;;) {                                                              \
+                long __int = prepare_to_wait_event(&wq_head, &__wq_entry, state);\
+                                                                                \
+                if (condition)                                                  \
+                        break;                                                  \
+		if (time_after_eq(jiffies, __end))				\
+			break;							\
+                                                                                \
+                if (___wait_is_interruptible(state) && __int) {                 \
+                        __ret = __int;                                          \
+                        goto __out;                                             \
+                }                                                               \
+                                                                                \
+                cmd;                                                            \
+        }                                                                       \
+        finish_wait(&wq_head, &__wq_entry);                                     \
+__out:  __ret;                                                                  \
+})
+
+#define __mt76_wait_event_timeout(wq_head, condition, timeout, exclusive)	\
+	___mt76_wait_event(wq_head, ___wait_cond_timeout(condition),		\
+		      TASK_UNINTERRUPTIBLE, exclusive, timeout,			\
+		      cond_resched())
+
+#define mt76_wait_event_timeout(wq_head, condition, timeout, exclusive)         \
+({                                                                              \
+        long __ret = timeout;                                                   \
+        might_sleep();                                                          \
+        if (!___wait_cond_timeout(condition))                                   \
+                __ret = __mt76_wait_event_timeout(wq_head, condition, timeout, exclusive); \
+        __ret;                                                                  \
+})
 
 struct sk_buff *
 __mt76_mcu_msg_alloc(struct mt76_dev *dev, const void *data,
@@ -30,22 +93,35 @@
 EXPORT_SYMBOL_GPL(__mt76_mcu_msg_alloc);
 
 struct sk_buff *mt76_mcu_get_response(struct mt76_dev *dev,
-				      unsigned long expires)
+				      unsigned long expires, int cmd)
 {
 	unsigned long timeout;
+	int is_ext_cmd = ((cmd & __MCU_CMD_FIELD_ID) == MCU_CMD_EXT_CID);
 
 	if (!time_is_after_jiffies(expires))
 		return NULL;
 
 	timeout = expires - jiffies;
-	wait_event_timeout(dev->mcu.wait,
+	mt76_wait_event_timeout(dev->mcu.wait,
 			   (!skb_queue_empty(&dev->mcu.res_q) ||
 			    test_bit(MT76_MCU_RESET, &dev->phy.state)),
-			   timeout);
+			   timeout, is_ext_cmd);
 	return skb_dequeue(&dev->mcu.res_q);
 }
 EXPORT_SYMBOL_GPL(mt76_mcu_get_response);
 
+bool mt76_get_test_mcu_restart(void)
+{
+	return rx_poll_retry_enable;
+}
+EXPORT_SYMBOL_GPL(mt76_get_test_mcu_restart);
+
+void mt76_set_test_mcu_restart(bool ok)
+{
+	rx_poll_retry_enable = ok;
+}
+EXPORT_SYMBOL_GPL(mt76_set_test_mcu_restart);
+
 void mt76_mcu_rx_event(struct mt76_dev *dev, struct sk_buff *skb)
 {
 	skb_queue_tail(&dev->mcu.res_q, skb);
@@ -81,6 +157,14 @@
 
 	mutex_lock(&dev->mcu.mutex);
 
+	//if (MCU_EXT_CMD_GET_MIB_INFO == (FIELD_GET(__MCU_CMD_FIELD_EXT_ID, cmd)) && 
+	if (dev->mcu.mib_access_time > 0 && 
+			time_after(dev->mcu.mib_access_time + HZ/10, jiffies)) {
+		ret = -EBUSY; 
+		dev_kfree_skb(skb);
+		goto out;
+	}
+
 	ret = dev->mcu_ops->mcu_skb_send_msg(dev, skb, cmd, &seq);
 	if (ret < 0) {
 		dev_kfree_skb(skb);
@@ -92,18 +176,23 @@
 		goto out;
 	}
 
+	if (dev->mcu_ops->mcu_set_timeout)
+		dev->mcu_ops->mcu_set_timeout(dev, cmd);
+
 	expires = jiffies + dev->mcu.timeout;
 
 	do {
-		skb = mt76_mcu_get_response(dev, expires);
+		skb = mt76_mcu_get_response(dev, expires, cmd);
 		ret = dev->mcu_ops->mcu_parse_response(dev, cmd, skb, seq);
 		if (!ret && ret_skb)
 			*ret_skb = skb;
 		else
 			dev_kfree_skb(skb);
-	} while (ret == -EAGAIN);
+	} while (ret == -EAGAIN && cmd != MCU_EXT_CMD(GET_MIB_INFO));
 
 out:
+//	if (MCU_EXT_CMD_GET_MIB_INFO == (FIELD_GET(__MCU_CMD_FIELD_EXT_ID, cmd)))
+		dev->mcu.mib_access_time = jiffies; // save the last MIB access time
 	mutex_unlock(&dev->mcu.mutex);
 
 	return ret;
--- a/mt7915/mcu.c	2026-06-27 22:32:12.657558320 +0800
+++ b/mt7915/mcu.c	2026-06-27 22:27:00.720376601 +0800
@@ -150,6 +150,7 @@
 		ht_mcs[nss] = sta->deflink.ht_cap.mcs.rx_mask[nss] & mask[nss];
 }
 
+static int mt7915_firmware_state(struct mt7915_dev *dev, bool wa);
 static int
 mt7915_mcu_parse_response(struct mt76_dev *mdev, int cmd,
 			  struct sk_buff *skb, int seq)
@@ -183,7 +184,17 @@
 	} else {
 		skb_pull(skb, sizeof(struct mt76_connac2_mcu_rxd));
 	}
-
+/* after calling __mt76_mcu_restart(mdev), the whole system will be shutdown?
+	if (mt76_get_test_mcu_restart()) {
+               if (!mt7915_firmware_state(dev, true)) { // firmware must be ready !
+			skb_queue_purge(&mdev->mcu.res_q);
+			dev_err(mdev->dev, "====== skb_queue_purge");
+			__mt76_mcu_restart(mdev);
+			dev_err(mdev->dev, "====== __mt76_mcu_restart");
+                }
+	       mt76_set_test_mcu_restart(false);
+	}
+*/
 	return ret;
 }
 
@@ -2466,6 +2477,7 @@
 		.mcu_skb_send_msg = mt7915_mcu_send_message,
 		.mcu_parse_response = mt7915_mcu_parse_response,
 		.mcu_restart = mt76_connac_mcu_restart,
+		.mcu_set_timeout = mt7915_mcu_set_timeout,
 	};
 
 	dev->mt76.mcu_ops = &mt7915_mcu_ops;
