--- a/mt76.h	2026-07-05 14:00:06.492715935 +0800
+++ b/mt76.h	2026-07-03 19:21:01.348728975 +0800
@@ -7,6 +7,8 @@
 #define __MT76_H
 
 #include <linux/kernel.h>
+#include <linux/wait.h>
+#include <linux/rcupdate.h>
 #include <linux/io.h>
 #include <linux/spinlock.h>
 #include <linux/skbuff.h>
@@ -223,6 +225,9 @@
 	int (*mcu_rd_rp)(struct mt76_dev *dev, u32 base,
 			 struct mt76_reg_pair *rp, int len);
 	int (*mcu_restart)(struct mt76_dev *dev);
+	void (*mcu_set_timeout)(struct mt76_dev *dev, int cmd);
+	bool (*mcu_limit_rate)(struct mt76_dev *dev, int cmd);
+	void (*mcu_access_time_update)(struct mt76_dev *dev, int cmd);
 };
 
 struct mt76_queue_ops {
@@ -520,13 +525,59 @@
 	__MT_EP_OUT_MAX,
 };
 
+#define ___mt76_wait_event(wq_head, condition, state, exclusive, ret, cmd)      \
+({                                                                              \
+        __label__ __out;                                                        \
+        struct wait_queue_entry __wq_entry;                                     \
+        long __ret = ret;       /* explicit shadow */                           \
+        unsigned long __end = jiffies + (ret);                                  \
+	unsigned int wq_flags = (exclusive || (cmd == MCU_EXT_CMD(GET_MIB_INFO))) ? WQ_FLAG_EXCLUSIVE : 0; \
+                                                                                \
+        init_wait_entry(&__wq_entry, wq_flags);                                 \
+        for (;;) {                                                              \
+                long __int = prepare_to_wait_event(&wq_head, &__wq_entry, state);\
+                                                                                \
+                if (condition)                                                  \
+                        break;                                                  \
+                if (time_after_eq(jiffies, __end))                              \
+                        break;                                                  \
+                                                                                \
+                if (___wait_is_interruptible(state) && __int) {                 \
+                        __ret = __int;                                          \
+                        goto __out;                                             \
+                }                                                               \
+                                                                                \
+                cmd;                                                            \
+        }                                                                       \
+        finish_wait(&wq_head, &__wq_entry);                                     \
+	smp_mb(); rcu_all_qs();							\
+__out:  __ret;                                                                  \
+})
+
+#define __mt76_wait_event_timeout(wq_head, condition, timeout, exclusive)       \
+        ___mt76_wait_event(wq_head, ___wait_cond_timeout(condition),            \
+                      TASK_UNINTERRUPTIBLE, exclusive, timeout,                 \
+                      cond_resched())
+
+#define mt76_wait_event_timeout(wq_head, condition, timeout, exclusive)         \
+({                                                                              \
+        long __ret = timeout;                                                   \
+        might_sleep();                                                          \
+        if (!___wait_cond_timeout(condition))                                   \
+                __ret = __mt76_wait_event_timeout(wq_head, condition, timeout, exclusive); \
+        __ret;                                                                  \
+})
+
 struct mt76_mcu {
 	struct mutex mutex;
 	u32 msg_seq;
 	int timeout;
+	unsigned long last_access_time;
+        unsigned long mib_access_time;
 
 	struct sk_buff_head res_q;
 	wait_queue_head_t wait;
+	wait_queue_head_t delay;
 };
 
 #define MT_TX_SG_MAX_SIZE	8
@@ -1390,7 +1441,9 @@
 
 void mt76_mcu_rx_event(struct mt76_dev *dev, struct sk_buff *skb);
 struct sk_buff *mt76_mcu_get_response(struct mt76_dev *dev,
-				      unsigned long expires);
+				      unsigned long expires, int cmd);
+bool mt76_get_test_mcu_restart(void);
+void mt76_set_test_mcu_restart(bool ok);
 int mt76_mcu_send_and_get_msg(struct mt76_dev *dev, int cmd, const void *data,
 			      int len, bool wait_resp, struct sk_buff **ret);
 int mt76_mcu_skb_send_and_get_msg(struct mt76_dev *dev, struct sk_buff *skb,
--- a/mcu.c	2026-07-05 14:00:06.448714079 +0800
+++ b/mcu.c	2026-07-05 14:01:06.763265734 +0800
@@ -4,6 +4,25 @@
  */
 
 #include "mt76.h"
+#include "mt76_connac_mcu.h"
+
+/*
+#define mt76_busy_wait_event_timeout(wq_head, condition, timeout)	\
+do {									\
+	unsigned long __last_qs = jiffies;				\
+	unsigned long __end = __last_qs + (timeout);			\
+	for (;;) {							\
+		rmb();							\
+		if (condition)						\
+			break;						\
+		if (time_after_eq(jiffies, __end))			\
+			break;						\
+		cond_resched();						\
+		msleep(1); cpu_relax();				\
+		rcu_softirq_qs_periodic(__last_qs);			\
+	}								\
+} while (0)
+*/
 
 struct sk_buff *
 __mt76_mcu_msg_alloc(struct mt76_dev *dev, const void *data,
@@ -30,22 +49,35 @@
 EXPORT_SYMBOL_GPL(__mt76_mcu_msg_alloc);
 
 struct sk_buff *mt76_mcu_get_response(struct mt76_dev *dev,
-				      unsigned long expires)
+				      unsigned long expires, int cmd)
 {
 	unsigned long timeout;
+	int is_ext_cmd = ((cmd & __MCU_CMD_FIELD_ID) == MCU_CMD_EXT_CID);
 
 	if (!time_is_after_jiffies(expires))
 		return NULL;
 
 	timeout = expires - jiffies;
-	wait_event_timeout(dev->mcu.wait,
+	mt76_wait_event_timeout(dev->mcu.wait,
 			   (!skb_queue_empty(&dev->mcu.res_q) ||
 			    test_bit(MT76_MCU_RESET, &dev->phy.state)),
-			   timeout);
+			   timeout, is_ext_cmd);
 	return skb_dequeue(&dev->mcu.res_q);
 }
 EXPORT_SYMBOL_GPL(mt76_mcu_get_response);
 
+bool mt76_get_test_mcu_restart(void)
+{
+	return rx_poll_retry_enable;
+}
+EXPORT_SYMBOL_GPL(mt76_get_test_mcu_restart);
+
+void mt76_set_test_mcu_restart(bool ok)
+{
+	rx_poll_retry_enable = ok;
+}
+EXPORT_SYMBOL_GPL(mt76_set_test_mcu_restart);
+
 void mt76_mcu_rx_event(struct mt76_dev *dev, struct sk_buff *skb)
 {
 	skb_queue_tail(&dev->mcu.res_q, skb);
@@ -79,7 +111,15 @@
 	if (ret_skb)
 		*ret_skb = NULL;
 
-	mutex_lock(&dev->mcu.mutex);
+        mutex_lock(&dev->mcu.mutex);
+
+#ifdef CONFIG_MIPS
+	smp_rmb();
+	if (unlikely(!skb)) {
+		dev_err(dev->dev, "Invalid skb\n");
+		goto out;
+	}
+#endif
 
 	ret = dev->mcu_ops->mcu_skb_send_msg(dev, skb, cmd, &seq);
 	if (ret < 0) {
@@ -92,18 +132,28 @@
 		goto out;
 	}
 
+	if (dev->mcu_ops->mcu_set_timeout)
+		dev->mcu_ops->mcu_set_timeout(dev, cmd);
+
 	expires = jiffies + dev->mcu.timeout;
 
 	do {
-		skb = mt76_mcu_get_response(dev, expires);
+		skb = mt76_mcu_get_response(dev, expires, cmd);
 		ret = dev->mcu_ops->mcu_parse_response(dev, cmd, skb, seq);
 		if (!ret && ret_skb)
 			*ret_skb = skb;
 		else
 			dev_kfree_skb(skb);
-	} while (ret == -EAGAIN);
+	} while (ret == -EAGAIN && cmd != MCU_EXT_CMD(GET_MIB_INFO));
 
 out:
+	if (dev->mcu_ops->mcu_limit_rate) {
+		dev->mcu_ops->mcu_limit_rate(dev, cmd);
+	}
+	
+        if (dev->mcu_ops->mcu_access_time_update)
+                dev->mcu_ops->mcu_access_time_update(dev, cmd);
+
 	mutex_unlock(&dev->mcu.mutex);
 
 	return ret;
--- a/mt7915/mcu.c	2026-07-05 14:00:06.492715935 +0800
+++ b/mt7915/mcu.c	2026-07-05 01:33:08.197875055 +0800
@@ -2,6 +2,8 @@
 /* Copyright (C) 2020 MediaTek Inc. */
 
 #include <linux/fs.h>
+#include <linux/nmi.h>
+#include <linux/wait.h>
 #include "mt7915.h"
 #include "mcu.h"
 #include "mac.h"
@@ -150,6 +152,7 @@
 		ht_mcs[nss] = sta->deflink.ht_cap.mcs.rx_mask[nss] & mask[nss];
 }
 
+static int mt7915_firmware_state(struct mt7915_dev *dev, bool wa);
 static int
 mt7915_mcu_parse_response(struct mt76_dev *mdev, int cmd,
 			  struct sk_buff *skb, int seq)
@@ -167,6 +170,8 @@
 		wake_up(&dev->mt76.mcu.wait);
 		queue_work(dev->mt76.wq, &dev->reset_work);
 		wake_up(&dev->reset_wait);*/
+		dev_err(mdev->dev, "MT7915: MCU 120s timeout! Firmware is completely dead. Panic now!\n");
+		mdelay(50); panic("mt7915: Crucial MCU command %08x timed out for 120s. System forced restart.\n", cmd);
 		return -ETIMEDOUT;
 	}
 
@@ -183,7 +188,17 @@
 	} else {
 		skb_pull(skb, sizeof(struct mt76_connac2_mcu_rxd));
 	}
-
+/* after calling __mt76_mcu_restart(mdev), the whole system will be shutdown?
+	if (mt76_get_test_mcu_restart()) {
+               if (!mt7915_firmware_state(dev, true)) { // firmware must be ready !
+			skb_queue_purge(&mdev->mcu.res_q);
+			dev_err(mdev->dev, "====== skb_queue_purge");
+			__mt76_mcu_restart(mdev);
+			dev_err(mdev->dev, "====== __mt76_mcu_restart");
+                }
+	       mt76_set_test_mcu_restart(false);
+	}
+*/
 	return ret;
 }
 
@@ -192,7 +207,7 @@
 static void
 mt7915_mcu_set_timeout(struct mt76_dev *mdev, int cmd)
 {
-	mdev->mcu.timeout = 5 * HZ;
+	mdev->mcu.timeout = 300 * HZ;
 
 	if ((cmd & __MCU_CMD_FIELD_ID) != MCU_CMD_EXT_CID)
 		return;
@@ -203,7 +218,7 @@
 	case MCU_EXT_CMD_PHY_STAT_INFO:
 	case MCU_EXT_CMD_STA_REC_UPDATE:
 	case MCU_EXT_CMD_BSS_INFO_UPDATE:
-		mdev->mcu.timeout = 2 * HZ;
+		mdev->mcu.timeout = 120 * HZ;
 		return;
 	default:
 		break;
@@ -2459,18 +2474,81 @@
 				 MCU_WA_PARAM_RED, 0, 0);
 }
 
+#define MCU_REFUSE_INTERVAL	HZ*2 /*2secs*/
+#define MCU_MIN_INTERVAL	HZ/10 /*100ms*/
+#define MCU_MAX_INTERVAL	HZ/5 /*200ms*/
+bool mt7915_mcu_limit_rate(struct mt76_dev *mdev, int cmd)
+{
+	bool ret = false;
+
+	struct mt7915_dev *dev = container_of(mdev, struct mt7915_dev, mt76);
+	if (dev && !dev->recovery.hw_init_done)
+		return ret;
+
+        if (mdev->mcu.last_access_time > 0 &&
+		(cmd != MCU_EXT_CMD(GET_MIB_INFO)) &&
+		(jiffies - mdev->mcu.last_access_time) < MCU_MIN_INTERVAL) {
+		unsigned long __end = jiffies + (MCU_MIN_INTERVAL - jiffies + mdev->mcu.last_access_time);
+#ifndef CONFIG_PREEMPTION
+		int sleep_interval = (MCU_MIN_INTERVAL < 1) ? MCU_MIN_INTERVAL : 1;
+		static unsigned long last_watchdog = 0;
+		for (;;) {
+			if ((jiffies - __end) >= 0)
+				break;
+
+			//udelay(sleep_us);
+			//cond_resched();
+			//set_current_state(TASK_INTERRUPTIBLE);
+			//schedule_timeout(usecs_to_jiffies(sleep_us)); 
+			smp_mb();
+			wait_event_timeout(mdev->mcu.delay, ((jiffies - __end) >= 0), sleep_interval);
+			cond_resched();
+		}
+		smp_mb();
+		cond_resched();
+		if (time_after(jiffies, last_watchdog+HZ)){
+			touch_nmi_watchdog();
+			last_watchdog = jiffies;
+		}
+#endif
+	}
+
+	if (mdev->mcu.mib_access_time > 0 && 
+		(cmd == MCU_EXT_CMD(GET_MIB_INFO)) &&
+		(jiffies - mdev->mcu.mib_access_time) < MCU_REFUSE_INTERVAL) {
+		ret = true;
+	}
+
+	return ret;
+}
+
+void mt7915_mcu_access_time_update(struct mt76_dev *mdev, int cmd)
+{
+	if (cmd == MCU_EXT_CMD(GET_MIB_INFO))
+	        mdev->mcu.mib_access_time = jiffies; // save the last MIB access time
+	else
+		mdev->mcu.last_access_time = jiffies;
+}
+
 int mt7915_mcu_init(struct mt7915_dev *dev)
 {
+	int ret;
+
 	static const struct mt76_mcu_ops mt7915_mcu_ops = {
 		.headroom = sizeof(struct mt76_connac2_mcu_txd),
 		.mcu_skb_send_msg = mt7915_mcu_send_message,
 		.mcu_parse_response = mt7915_mcu_parse_response,
 		.mcu_restart = mt76_connac_mcu_restart,
+		.mcu_set_timeout = mt7915_mcu_set_timeout,
+		.mcu_limit_rate = mt7915_mcu_limit_rate,
+		.mcu_access_time_update = mt7915_mcu_access_time_update,
 	};
 
 	dev->mt76.mcu_ops = &mt7915_mcu_ops;
 
-	return mt7915_mcu_init_firmware(dev);
+	ret = mt7915_mcu_init_firmware(dev);
+
+	return ret;
 }
 
 void mt7915_mcu_exit(struct mt7915_dev *dev)
@@ -3116,6 +3194,8 @@
 	int i, ret, len, offs_cc;
 	u64 cc_tx;
 
+        return 0; // firmware bug. return directly
+
 	/* strict order */
 	if (is_mt7915(&dev->mt76)) {
 		static const u32 chip_offs[] = {
--- a/mt7915/init.c	2026-07-05 14:00:06.488715766 +0800
+++ b/mt7915/init.c	2026-06-30 18:15:15.544887831 +0800
@@ -415,6 +415,8 @@
 				   "mediatek,disable-radar-background"))
 		wiphy_ext_feature_set(wiphy,
 				      NL80211_EXT_FEATURE_RADAR_BACKGROUND);
+	else
+		dev_err(mdev->dev, "====== MT7915: mediatek,disable-radar-background @%s\n", __func__);
 
 	/*if (profile_has_rate_control(mdev->dev)) {
 		ieee80211_hw_set(hw, HAS_RATE_CONTROL);
