--- a/mt76.h	2026-08-04 08:54:00.021134299 +0800
+++ b/mt76.h	2026-08-04 13:52:55.304259656 +0800
@@ -525,6 +525,14 @@
 	__MT_EP_OUT_MAX,
 };
 
+#define mt76_delay_yield(iterations) do { \
+    int __i; \
+    for (__i = 0; __i < (iterations); __i++) { \
+        udelay(1); \
+        cond_resched(); \
+    } \
+} while (0)
+
 #define ___mt76_wait_event(wq_head, condition, state, exclusive, ret, cmd)      \
 ({                                                                              \
         __label__ __out;                                                        \
--- a/dma.c	2026-08-04 08:54:00.013133735 +0800
+++ b/dma.c	2026-08-04 13:54:06.646574777 +0800
@@ -964,11 +964,11 @@
 		if (done) wmb(); // sync. cpu write
 		dev->drv->rx_poll_complete(dev, qid);
 		if (rx_poll_retry_enable) mt76_rx_poll_retry_adjust(start_jif);
-		udelay(10); return done;
+		mt76_delay_yield(10); return done;
 	}else if (--retry_limit >= 0) {
 		if (rx_poll_retry_enable) mt76_rx_poll_retry_adjust(start_jif);
 		done = 0;
-		udelay(10); goto rev_dma;
+		mt76_delay_yield(10); goto rev_dma;
 	}
 
 	if (rx_poll_retry_enable) mt76_rx_poll_retry_adjust(start_jif);
