--- a/dma.c	2026-09-03 09:38:22.328955574 +0800
+++ b/dma.c	2026-09-03 09:37:13.034355413 +0800
@@ -599,6 +599,7 @@
 		struct mt76_queue_buf qbuf;
 		void *buf = NULL;
 
+		smp_mb();
 		buf = page_frag_alloc(&q->rx_page, q->buf_size, GFP_ATOMIC|GFP_DMA);
 		if (!buf)
 			break;
@@ -946,7 +947,17 @@
 	int retry_limit = (!rx_poll_retry_enable) ? 0 : atomic_read(&rx_poll_retry_cnt); 
 	u64 start_jif = ktime_get();
 
+	if (unlikely(!napi || !virt_addr_valid(napi) || !virt_addr_valid(napi->dev))) {
+		pr_err_ratelimited("mt76: FATAL! napi or napi->dev pointer is invalid VA!\n");
+		return 0;
+	}
+
 	dev = container_of(napi->dev, struct mt76_dev, napi_dev);
+	if (unlikely(!virt_addr_valid(dev))) {
+		pr_err_ratelimited("mt76: FATAL! dev pointer is invalid VA: %p, napi: %p\n", dev, napi);
+		return 0;
+	}
+
 	qid = napi - dev->napi;
 
 rev_dma:
