--- a/mt76.h
+++ b/mt76.h
@@ -15,6 +15,9 @@
 #include <linux/average.h>
 #include <linux/soc/mediatek/mtk_wed.h>
 #include <net/mac80211.h>
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+#include <net/page_pool.h>
+#endif
 #include "util.h"
 #include "testmode.h"
 
@@ -204,6 +207,9 @@
 	dma_addr_t desc_dma;
 	struct sk_buff *rx_head;
 	struct page_frag_cache rx_page;
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+	struct page_pool *page_pool;
+#endif
 };
 
 struct mt76_mcu_ops {
@@ -352,6 +358,9 @@
 		struct sk_buff *skb;
 		void *ptr;
 	};
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+	struct page_pool *pp; 
+#endif
 };
 
 struct mt76_rx_tid {
--- a/dma.c
+++ b/dma.c
@@ -172,8 +172,14 @@
 
 	local_bh_disable();
 	while ((t = __mt76_get_rxwi(dev)) != NULL) {
-		if (t->ptr)
+		if (t->ptr) {
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+			if (t->pp)
+				page_pool_put_page(t->pp, virt_to_head_page(t->ptr), false);
+			else
+#endif
 			skb_free_frag(t->ptr);
+		}
 		kfree(t);
 	}
 	local_bh_enable();
@@ -223,7 +229,9 @@
 		txwi = mt76_get_rxwi(dev);
 		if (!txwi)
 			return -ENOMEM;
-
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+		txwi->pp = q->page_pool;
+#endif
 		rx_token = mt76_rx_token_consume(dev, data, txwi, buf->addr);
 		if (rx_token < 0) {
 			mt76_put_rxwi(dev, txwi);
@@ -410,10 +418,17 @@
 		if (!t)
 			return NULL;
 
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+		if (!t->pp) {
+	                dma_unmap_single(dev->dma_dev, t->dma_addr,
+        	                         SKB_WITH_OVERHEAD(q->buf_size),
+                	                 DMA_FROM_DEVICE);
+		}
+#else
 		dma_unmap_single(dev->dma_dev, t->dma_addr,
 				 SKB_WITH_OVERHEAD(q->buf_size),
 				 DMA_FROM_DEVICE);
-
+#endif
 		buf = t->ptr;
 		t->dma_addr = 0;
 		t->ptr = NULL;
@@ -427,11 +442,22 @@
 					   MT_DMA_CTL_DROP));
 		}
 	} else {
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+		struct mt76_txwi_cache *t = e->txwi;
+#endif
 		buf = e->buf;
 		e->buf = NULL;
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+		if (!t || !t->pp) {
+			dma_unmap_single(dev->dma_dev, e->dma_addr[0],
+					 SKB_WITH_OVERHEAD(q->buf_size),
+					 DMA_FROM_DEVICE);
+		}
+#else
 		dma_unmap_single(dev->dma_dev, e->dma_addr[0],
 				 SKB_WITH_OVERHEAD(q->buf_size),
 				 DMA_FROM_DEVICE);
+#endif
 	}
 
 	return buf;
@@ -586,7 +612,7 @@
 mt76_dma_rx_fill(struct mt76_dev *dev, struct mt76_queue *q)
 {
 	int len = SKB_WITH_OVERHEAD(q->buf_size);
-	int frames = 0, offset = q->buf_offset;
+	int frames = 0, offset;
 	dma_addr_t addr;
 
 	if (!q->ndesc)
@@ -598,23 +624,50 @@
 		struct mt76_queue_buf qbuf;
 		void *buf = NULL;
 
-		buf = page_frag_alloc(&q->rx_page, q->buf_size, GFP_ATOMIC|GFP_DMA);
-		if (!buf)
-			break;
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+		if (q->page_pool) {
+			struct page *page;
+
+			page = page_pool_alloc_frag(q->page_pool, &offset,
+						   q->buf_size, GFP_ATOMIC | __GFP_NOWARN | GFP_DMA);
+			if (!page)
+				break;
 
-		addr = dma_map_single(dev->dma_dev, buf, len, DMA_FROM_DEVICE);
-		if (unlikely(dma_mapping_error(dev->dma_dev, addr))) {
-			skb_free_frag(buf);
-			break;
+			buf = page_address(page) + offset; // virtual address
+
+			addr = page_pool_get_dma_addr(virt_to_head_page(buf)) + offset; // physical address
+
+			qbuf.skip_unmap = true;
+		} else
+#endif
+		{
+			buf = page_frag_alloc(&q->rx_page, q->buf_size, GFP_ATOMIC|GFP_DMA);
+			if (!buf)
+				break;
+
+			addr = dma_map_single(dev->dma_dev, buf, len, DMA_FROM_DEVICE);
+			if (unlikely(dma_mapping_error(dev->dma_dev, addr))) {
+				skb_free_frag(buf);
+				break;
+			}
+
+			qbuf.skip_unmap = false;
 		}
 
-		qbuf.addr = addr + offset;
-		qbuf.len = len - offset;
-		qbuf.skip_unmap = false;
+		qbuf.addr = addr + q->buf_offset;
+                qbuf.len = len - q->buf_offset;
+
 		if (mt76_dma_add_rx_buf(dev, q, &qbuf, buf) < 0) {
-			dma_unmap_single(dev->dma_dev, addr, len,
-					 DMA_FROM_DEVICE);
-			skb_free_frag(buf);
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+			if (q->page_pool) {
+				page_pool_put_page(q->page_pool, virt_to_head_page(buf), false);
+			} else
+#endif
+			{
+				dma_unmap_single(dev->dma_dev, addr, len,
+						 DMA_FROM_DEVICE);
+				skb_free_frag(buf);
+			}
 			break;
 		}
 		frames++;
@@ -727,7 +780,11 @@
 		buf = mt76_dma_dequeue(dev, q, true, NULL, NULL, &more, NULL);
 		if (!buf)
 			break;
-
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+                if (q->page_pool)
+                        page_pool_put_page(q->page_pool, virt_to_head_page(buf), false);
+                else
+#endif
 		skb_free_frag(buf);
 	} while (1);
 	spin_unlock_bh(&q->lock);
@@ -777,6 +834,11 @@
 
 		skb_add_rx_frag(skb, nr_frags, page, offset, len, q->buf_size);
 	} else {
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+		if (q->page_pool)
+			page_pool_put_page(q->page_pool, virt_to_head_page(data), false);
+		else
+#endif
 		skb_free_frag(data);
 	}
 
@@ -822,6 +884,15 @@
 		if (!data)
 			break;
 
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+		if (q->page_pool) {
+			struct page *page = virt_to_head_page(data);
+			dma_addr_t dma_addr = page_pool_get_dma_addr(page);
+
+			dma_sync_single_for_cpu(dev->dma_dev, dma_addr, len, DMA_FROM_DEVICE);
+		}
+#endif
+
 		if (drop)
 			goto free_frag;
 
@@ -865,6 +936,11 @@
 		continue;
 
 free_frag:
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+		if (q->page_pool)
+			page_pool_put_page(q->page_pool, virt_to_head_page(data), false);
+		else
+#endif
 		skb_free_frag(data);
 	}
 
@@ -910,7 +986,26 @@
 	dev->napi_dev.threaded = 1;
 
 	mt76_for_each_q_rx(dev, i) {
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+		struct mt76_queue *q = &dev->q_rx[i];
+#endif
 		netif_napi_add(&dev->napi_dev, &dev->napi[i], poll);
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+		if (!q->page_pool) {
+			struct page_pool_params pp_params = {
+			    .flags = PP_FLAG_DMA_MAP,
+			    .order = 0,
+			    .pool_size = q->ndesc,
+			    .nid = NUMA_NO_NODE,
+			    .dev = dev->dma_dev,
+			    .dma_dir = DMA_FROM_DEVICE,
+			};
+			q->page_pool = page_pool_create(&pp_params);
+		}
+		if (IS_ERR(q->page_pool)) {
+			q->page_pool = NULL;
+		}
+#endif
 		mt76_dma_rx_fill(dev, &dev->q_rx[i]);
 		napi_enable(&dev->napi[i]);
 	}
@@ -957,17 +1052,27 @@
 	for (i = 0; i < ARRAY_SIZE(dev->q_mcu); i++)
 		mt76_dma_tx_cleanup(dev, dev->q_mcu[i], true);
 
+        mt76_free_pending_txwi(dev);
+        mt76_free_pending_rxwi(dev);
+
 	mt76_for_each_q_rx(dev, i) {
 		struct mt76_queue *q = &dev->q_rx[i];
 
 		netif_napi_del(&dev->napi[i]);
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+		if (FIELD_GET(MT_QFLAG_WED_TYPE, q->flags) || q->page_pool)
+#else
 		if (FIELD_GET(MT_QFLAG_WED_TYPE, q->flags))
+#endif
 			mt76_dma_rx_cleanup(dev, q);
+#if IS_ENABLED(CONFIG_PAGE_POOL)
+		if (q->page_pool) {
+			page_pool_destroy(q->page_pool);
+			q->page_pool = NULL;
+		}
+#endif
 	}
 
-	mt76_free_pending_txwi(dev);
-	mt76_free_pending_rxwi(dev);
-
 	if (mtk_wed_device_active(&dev->mmio.wed))
 		mtk_wed_device_detach(&dev->mmio.wed);
 }
